include/oneapi/dpl/internal/histogram_impl.h

```diff
@@ -0,0 +1,116 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_HISTOGRAM_IMPL_H
+#define _ONEDPL_HISTOGRAM_IMPL_H
+
+#include "function.h"
+#include "histogram_extension_defs.h"
+#include "histogram_binhash_utils.h"
+#include "../pstl/iterator_impl.h"
+
+#if _ONEDPL_HETERO_BACKEND
+#    include "oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h"
+#endif
+
+namespace oneapi
+{
+namespace dpl
+{
+
+namespace __internal
+{
+
+//Only supported for dpcpp backend
+#if _ONEDPL_BACKEND_SYCL
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _IdxHashFunc,
+          typename _RandomAccessIterator2>
+inline oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<_ExecutionPolicy>::type>
+__pattern_histogram(_ExecutionPolicy&& exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _IdxHashFunc __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
```

Comment:
I think this is what we want.  The semantics are that if elements don't fall into bins (if they are out of range), they are silently skipped.  I see this case as all input elements are skipped because they don't map to any existing bin.

I'd be open to some error message instead, but I think this makes sense.